oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java

```diff
@@ -38,24 +43,59 @@
 import org.apache.jackrabbit.oak.plugins.memory.StringPropertyState;
 import org.apache.jackrabbit.oak.plugins.value.Conversions;
 import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * PropertyState implementation with lazy parsing of the JSOP encoded value.
  */
 final class DocumentPropertyState implements PropertyState {
 
+    private static final Logger LOG = LoggerFactory.getLogger(DocumentPropertyState.class);
+
     private final DocumentNodeStore store;
 
     private final String name;
 
-    private final String value;
+    private String value;
 
     private PropertyState parsed;
+    private byte[] compressedValue;
+    private final Compression compression = Compression.GZIP;
+
+    private static final int DEFAULT_COMPRESSION_THRESHOLD = Integer.getInteger("oak.mongo.compressionThreshold", 1024);
 
     DocumentPropertyState(DocumentNodeStore store, String name, String value) {
         this.store = store;
         this.name = name;
-        this.value = value;
+
+        int size = value.getBytes().length;
+        if (size > DEFAULT_COMPRESSION_THRESHOLD) {
+            try {
+                compressedValue = compress(value.getBytes());
+                this.value = null;
+            } catch (Exception e) {
+                LOG.warn("Failed to compress property {} value: ", name, e);
+                this.value = value;
+                this.compressedValue = null;
+            }
+        } else {
+            this.value = value;
+            compressedValue = null;
+        }
+    }
+
+    private byte[] compress(byte[] value) {
+        try {
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
+            OutputStream compressionOutputStream = compression.getOutputStream(out);
+            compressionOutputStream.write(value);
+            compressionOutputStream.close();
+            return out.toByteArray();
+        } catch (IOException e) {
```

Comment:
I would suggest to create a test case with compression/decompression throwing exceptions